File: src/main/java/com/pandus/leetcode/solutions/google/RemoveKDigits.java
package com.pandus.leetcode.solutions.google;

public class RemoveKDigits {
    public String removeKdigits(String num, int k) {
        StringBuilder sb = new StringBuilder();
        int remove = k;
        for (char c : num.toCharArray()) {
            while (remove > 0 && !sb.isEmpty() && sb.charAt(sb.length() - 1) > c) {
                sb.deleteCharAt(sb.length() - 1);
                remove--;
            }
            sb.append(c);
        }
        sb.setLength(sb.length() - remove);
        int i = 0;
        while (i < sb.length() && sb.charAt(i) == '0') {
            i++;
        }
        String res = sb.substring(i);
        return res.isEmpty() ? "0" : res;
    }
}

File: src/main/kotlin/com/pandus/leetcode/solutions/daily/RemoveKDigits.kt
package com.pandus.leetcode.solutions.daily

import java.util.*


//Reference: https://leetcode.com/problems/remove-k-digits
class RemoveKDigits {
    fun removeKdigits(num: String, k: Int): String {
        var k = k
        val stack = LinkedList<Char>()

        for (digit in num.toCharArray()) {
            while (stack.size > 0 && k > 0 && stack.peekLast() > digit) {
                stack.removeLast()
                k -= 1
            }
            stack.addLast(digit)
        }

        repeat(k) {
            stack.removeLast()
        }

        val leadingZerosRemoved = stack.dropWhile { it == '0' }
        return if (leadingZerosRemoved.isEmpty()) "0" else leadingZerosRemoved.joinToString("")
    }
}
